### PR TITLE
Make a type with signatures not be a subtype of anyFunctionType

### LINKED ISSUE #3954: Strange tuple typing
```typescript
function f<T extends { "0": (p1: number) => number }>(p: T): T
{
	return p;
}

var v = f([x => x]); // v is typed as [{}]. Expected [number=>number]
```

### PR DESCRIPTION
**Pull request by JsonFreeman**
Fixes #3954.

Here is an example:
```ts
function f<T extends [(p1: number) => number]>(p: T): T {
    return p;
}

var v = f([x => x]);
```

Type argument inference processes arguments in up to two stages. The first stage avoids contextually typing function expressions, and gives those function expressions a wildcard type (in the compiler called anyFunctionType). The second stage contextually types the functions, and infers the correct type. However, in both stages, we add inference candidates. Which means that in the first stage, we may add an inference candidate that contains the anyFunctionType. This "poisons the pot", and masks the good inference.

The solution leverages the final stage of type argument inference, namely the computation of the common supertype among the inference candidates. We'd like the candidate that contains the anyFunctionType *not* to win. The way to make something not win is to make it not a supertype of the better thing. So the way I've fixed this is by adding a new subtype rule:

Let S be the source type and T be the target. If S has signatures and T is the anyFunctionType, then S is not a subtype of T. Note that it's still assignable, it's just not a subtype. This is sufficient to make the anyFunctionType not win over the real function type.

### CODE REVIEW

**Comment by danquirk** [comment]
New behavior is or is not according to the current spec?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -5041,10 +5041,24 @@ namespace ts {
...
+                // with signatures is never a subtype of anyFunctionType (but it is assignable). This makes the type with
+                // signatures win over anyFunctionType.
+                if (source === anyFunctionType || (relation === assignableRelation && target === anyFunctionType)) {
                     return Ternary.True;
                 }
+
                 let sourceSignatures = getSignaturesOfType(source, kind);
+                if (target === anyFunctionType) {
+                    Debug.assert(relation === subtypeRelation);
+                    if (sourceSignatures.length > 0) {
```
Comment (yuit) [inline src/compiler/checker.ts]: I don't understand this check

Comment (JsonFreeman) [inline src/compiler/checker.ts]: Ah, basically, if the source has no signatures, I think it's still okay for it to be a subtype of anyFunctionType. There is nothing that necessarily makes it better than anyFunctionType. The case I care about is when it *does* have signatures.

**Comment by ahejlsberg** [comment]
Wouldn't it be better to avoid making *any* inferences from a type that is or contains the wildcard function type? Right now we protect against the *is* case but not the *contains* case. It seems to me that once we make an inference that contains the wildcard type there's a chance it might escape. Your fix makes sure it doesn't happen when there's another inference in the mix, but what if there isn't?

**Comment by JsonFreeman** [comment]
I'll try bubbling up a flag to indicate that the type contains anyFunctionType, like we discussed.

**Comment by JsonFreeman** [comment]
Closing this in favor of #3986.

**Pull request closed without merging**